### PR TITLE
[codex] Gate Airflow daily load readiness with DB validation

### PR DESCRIPTION
**Pull request by hoonly01**
## Summary
- Add `post_aggregate_validate` after `gold_aggregate` in the Airflow DAG.
- Validate target-date DB readiness from persisted tables before considering the DAG successful.
- Keep zero fresh ingestion as warning/report only; fail on state transition, metadata mapping, orphan integrity, mart freshness/count, and mart quality issues.
- Document the readiness contract and follow-up dashboard scope.

## Verification
- `PYTHONPATH=. uv run pytest -q`
  - `351 passed, 8 warnings`
- `basedpyright --level error src/pipeline/post_aggregate_validation.py src/pipeline/steps.py src/pipeline/cli.py dags/financial_review_pipeline.py`
  - `0 errors, 0 warnings, 0 notes`
- `git diff --check`

## Not tested
- Deployed Airflow scheduler/webserver DAG run.
- Airflow Web UI confirmation in staging/production.

## Operational notes
- This repo's compose files do not define Airflow webserver/scheduler services, so Web UI verification should happen in the deployed or staging Airflow environment.
- Metabase/Grafana dashboards remain a follow-up item, not part of this PR.

## Release Notes

* **New Features**
  * Added post-aggregate validation step that automatically verifies data quality after aggregation, checking for freshness, processing state, mapping completeness, referential integrity, and consistency across data marts

* **Documentation**
  * Published operational readiness guide detailing validation policies, success and failure criteria, local reproduction methods, and dashboard follow-up actions

* **Tests**
  * Added validation test suite covering success and failure scenarios with comprehensive edge-case coverage

### CODE REVIEW

File: src/pipeline/post_aggregate_validation.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+
+    def _check_review_state(self, session, target_date: date) -> ValidationCheck:
+        stuck_rows = session.execute(
+            text(
+                """
+                SELECT processing_status::text AS status, COUNT(*) AS count
+                FROM review_master_index
+                WHERE DATE_TRUNC('day', review_created_at)::date = :target_date
+                  AND processing_status::text IN :stuck_statuses
+                GROUP BY processing_status
```
[the remark this answers is not in the document]
Comment (hoonly01) [reply src/pipeline/post_aggregate_validation.py]: Decision: implemented

Rationale: Must Fix. Daily Airflow DB readiness validation is on the DAG critical path; `DATE_TRUNC` on timestamp columns can prevent index-friendly filtering and make validation cost grow with table size.

Commit hash: 7a393fc8d84c95f29024e61bb30f3c188cd8b5dd

Concise summary: Replaced target-date `DATE_TRUNC(...)=:target_date` predicates with half-open `>= :target_date` and `< :target_date_end` parameters across the validator timestamp queries.

Validation performed:
- `PYTHONPATH=. uv run pytest tests/test_post_aggregate_validation.py tests/test_pipeline_cli_validation.py tests/test_airflow_dag_validation_wiring.py -q` → 11 passed
- `PYTHONPATH=. uv run pytest -q` → 351 passed
- `basedpyright --level error ...` → 0 errors
- `git diff --check` → passed

File: src/pipeline/post_aggregate_validation.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+                FROM review_master_index rmi
+                LEFT JOIN app_metadata am
+                  ON am.app_id = rmi.app_id
+                 AND am.service_id = rmi.service_id
+                 AND am.is_active IS TRUE
+                 AND (am.valid_from IS NULL OR am.valid_from <= :target_date)
+                 AND (am.valid_to IS NULL OR am.valid_to >= :target_date)
+                WHERE DATE_TRUNC('day', rmi.review_created_at)::date = :target_date
+                  AND rmi.processing_status::text = 'ANALYZED'
+                """
```
[the remark this answers is not in the document]
Comment (hoonly01) [reply src/pipeline/post_aggregate_validation.py:209]: Decision: implemented

Rationale: Optional, but reasonable, low-risk, and in current scope. Removing column-side enum-to-text casts avoids unnecessary casts and keeps enum comparisons index-friendly. I kept the status sets as finite SQL literals instead of introducing broader enum-array binding refactors.

Commit hash: 7a393fc8d84c95f29024e61bb30f3c188cd8b5dd

Concise summary: Replaced `status::text` / `processing_status::text` reads and predicates with direct enum-column comparisons for the validation queries.

Validation performed:
- `PYTHONPATH=. uv run pytest tests/test_post_aggregate_validation.py tests/test_pipeline_cli_validation.py tests/test_airflow_dag_validation_wiring.py -q` → 11 passed
- `PYTHONPATH=. uv run pytest -q` → 351 passed
- `basedpyright --level error ...` → 0 errors
- `git diff --check` → passed

File: src/pipeline/post_aggregate_validation.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+                LEFT JOIN review_action_analysis raa
+                  ON raa.review_id = t.review_id
+                LEFT JOIN app_reviews ar
+                  ON ar.platform_review_id = t.platform_review_id
+                LEFT JOIN reviews_preprocessed rp
+                  ON rp.review_id = t.review_id
+                LEFT JOIN (
+                    SELECT review_id, COUNT(*) AS aspect_count
+                    FROM review_aspects
+                    GROUP BY review_id
```
[the remark this answers is not in the document]
Comment (hoonly01) [reply src/pipeline/post_aggregate_validation.py]: Decision: implemented

Rationale: Must Fix. The orphan-integrity check runs in the daily Airflow gate; grouping all historical `review_aspects` rows for every target date can become a production performance bug as data grows.

Commit hash: 7a393fc8d84c95f29024e61bb30f3c188cd8b5dd

Concise summary: Restricted `aspect_counts` by joining `review_aspects` to the target-date CTE before grouping, so only target review IDs are aggregated.

Validation performed:
- `PYTHONPATH=. uv run pytest tests/test_post_aggregate_validation.py tests/test_pipeline_cli_validation.py tests/test_airflow_dag_validation_wiring.py -q` → 11 passed
- `PYTHONPATH=. uv run pytest -q` → 351 passed
- `basedpyright --level error ...` → 0 errors
- `git diff --check` → passed

**Comment by hoonly01** [comment]
@coderabbit CodeRabbit review follow-up:

- Decision: implemented — target-scoped `aspect_counts` aggregation.
  - Rationale: Must Fix. The daily DAG gate should not group all historical `review_aspects` rows.
  - Commit hash: 7a393fc8d84c95f29024e61bb30f3c188cd8b5dd
  - Summary: Joined `review_aspects` to the target CTE before grouping.
- Decision: implemented — half-open target-date predicates.
  - Rationale: Must Fix. `DATE_TRUNC` on timestamp columns is not index-friendly on the DAG critical path.
  - Commit hash: 7a393fc8d84c95f29024e61bb30f3c188cd8b5dd
  - Summary: Bound `target_date_end` and used `>= :target_date` / `< :target_date_end`.
- Decision: implemented — whitespace-insensitive DAG dependency assertion.
  - Rationale: Optional, low-risk, and in scope.
  - Commit hash: 7a393fc8d84c95f29024e61bb30f3c188cd8b5dd
  - Summary: Replaced exact newline/indent matching with a regex dependency check.
- Decision: implemented — unique CLI test `storage_path`.
  - Rationale: Optional, low-risk, and reduces cross-run collision risk.
  - Commit hash: 7a393fc8d84c95f29024e61bb30f3c188cd8b5dd
  - Summary: Added a UUID suffix to the failure-test storage key.
- Decision: implemented — cleanup-helper SQL hardening.
  - Rationale: Optional, low-risk, and within test cleanup scope.
  - Commit hash: 7a393fc8d84c95f29024e61bb30f3c188cd8b5dd
  - Summary: Added strict partition-name validation and replaced table-name interpolation loop with explicit DELETE statements.
- Decision: not implemented — docstring coverage warning / Finishing Touches generators.
  - Rationale: Not Needed for this PR. The warning is broad repo-level hygiene and the generator checkboxes are optional automation suggestions, not required for the Airflow readiness gate.

Validation performed:
- `PYTHONPATH=. uv run pytest tests/test_post_aggregate_validation.py tests/test_pipeline_cli_validation.py tests/test_airflow_dag_validation_wiring.py -q` → 11 passed
- `PYTHONPATH=. uv run pytest -q` → 351 passed
- `basedpyright --level error src/pipeline/post_aggregate_validation.py tests/test_airflow_dag_validation_wiring.py tests/test_pipeline_cli_validation.py tests/test_post_aggregate_validation.py` → 0 errors
- `git diff --check` → passed
